False positive: let_var_whitespace with Swift 5.1 property wrappers

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

I have adopted the new Swift 5.1 property wrapper feature. It seems to cause a false positive with the opt-in rule `let_var_whitespace`.

Relevant file:
```
import Foundation

public enum Defaults {

    @UserDefault("isFirstTimeUser", defaultValue: true)
    public static var isFirstTimeUser: Bool
}
```

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint --config ../.swiftlint.yml 
Loading configuration from '../.swiftlint.yml'
Linting Swift files at paths 
Linting 'Defaults.swift' (1/1)
<path>/Defaults.swift:6:5: warning: Variable Declaration Whitespace Violation: Let and var should be separated from other statements by a blank line. (let_var_whitespace)
Done linting! Found 1 violation, 0 serious in 1 files.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 0.37.0
* Installation method used (Homebrew, CocoaPods, building from source, etc)? Homebrew
* Paste your configuration file:

```yml
opt_in_rules:
  - let_var_whitespace
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? No
* Which Xcode version are you using? 11.2.1
* Do you have a sample that shows the issue? 
```
echo "public enum Defaults {
@UserDefault("isFirstTimeUser", defaultValue: true)
public static var isFirstTimeUser: Bool
}" | swiftlint lint --no-cache --use-stdin --enable-all-rules
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -51,6 +51,11 @@
 * Error by default on bad expiring todo date formatting.  
   [Christopher Hale](https://github.com/chrispomeroyhale)
   [#3636](https://github.com/realm/SwiftLint/pull/3626)
+  
+* Avoid false-positive in `let_var_whitespace` rule by allowing custom
+  attributes on lines directly before let/var declarations.  
+  [SimplyDanny](https://github.com/SimplyDanny)
+  [#2980](https://github.com/realm/SwiftLint/issues/2980)
 
 ## 0.47.0: Smart Appliance
 

--- a/Source/SwiftLintFramework/Rules/Style/LetVarWhitespaceRule.swift
+++ b/Source/SwiftLintFramework/Rules/Style/LetVarWhitespaceRule.swift
@@ -32,6 +32,14 @@ public struct LetVarWhitespaceRule: ConfigurationProviderRule, OptInRule, Automa
 
             let x = bar as! Bar
             """),
+            Example("""
+                @available(swift 4)
+                @UserDefault("param", defaultValue: true)
+                var isEnabled = true
+
+                @Attribute
+                func f() {}
+            """),
             Example("var x: Int {\n\tlet a = 0\n\treturn a\n}\n") // don't trigger on local vars
         ],
         triggeringExamples: [
@@ -40,7 +48,17 @@ public struct LetVarWhitespaceRule: ConfigurationProviderRule, OptInRule, Automa
             Example("struct X {\n\tlet a\n\t↓func x() {}\n}\n"),
             Example("var x = 0\n↓@objc func f() {}\n"),
             Example("var x = 0\n↓@objc\n\tfunc f() {}\n"),
-            Example("@objc func f() {\n}\n↓var x = 0\n")
+            Example("@objc func f() {\n}\n↓var x = 0\n"),
+            Example("""
+                struct S {
+                    func f() {}
+                    ↓@Wapper
+                    let isNumber = false
+                    @Wapper
+                    var isEnabled = true
+                    ↓func g() {}
+                }
+            """)
         ]
     )
 
@@ -214,12 +232,22 @@ public struct LetVarWhitespaceRule: ConfigurationProviderRule, OptInRule, Automa
     // Collects all the line numbers containing attributes but not declarations
     // other than let/var
     private func attributeLineNumbers(file: SwiftLintFile) -> Set<Int> {
-        return Set(file.syntaxMap.tokens.compactMap({ token in
-            if token.kind == .attributeBuiltin {
-                return file.line(byteOffset: token.offset)
-            }
-            return nil
-        }))
+        let lineNumbers = file.syntaxMap.tokens
+            .filter { isAttribute(token: $0, in: file) }
+            .map(\.offset)
+            .compactMap(file.line)
+        return Set(lineNumbers)
+    }
+
+    private func isAttribute(token: SwiftLintSyntaxToken, in file: SwiftLintFile) -> Bool {
+        let kind = token.kind
+        if kind == .attributeBuiltin {
+            return true
+        }
+        if kind == .typeidentifier, let symbol = file.stringView.substringStartingLinesWithByteRange(token.range) {
+            return symbol.trimmingCharacters(in: .whitespaces).starts(with: "@")
+        }
+        return false
     }
 }
 

